Develop to Master (workshop session 2)

derekquinn opened the pull request:
_Thanks for stopping by, this work represents one day of the two-day React workshop I recently attended. I am looking for feedback to refine my understanding of basic concepts, so I can move on to more advanced techniques within the framework._ 
- Added TimeEntry
- Improved TimeLog
- Two-way binding with user input in Weather Component
- Added helper functions to WeatherForm
- Added WeatherList Component
- GradeForm functionality added
- Fixed bug in TimeLog delete button
- Refactored Messages component into a functional component

glidej commented on src/components/GradeForm.js:
```diff
@@ -0,0 +1,82 @@
+    
+        // return an object that gives new value (to set state)
+        this.setState(prev => {
+          const newList = prev.grades.slice(0);
```
This works perfectly fine, so if you're happy with it there's no need to change it. 

I have transitioned to using spread operators to do this work because creating a copy of the array/pushing a new element into it seemed like I was "juggling" variables. Here's how it could work with spread:

```
this.setState(currentState => {
    return {
        ...currentState, // first, "spread" the currentState into the new object literal. If we did nothing else, this would return an exact copy of `currentState` without the need to slice/copy
        grades: [ // lets use the same spread principle _inside_ the object to create a new `grades` array.
            ...currentState.grades, // pull in the current grades array and spread it in place
            { // merge in a new object to the end of the array
                label: this.state.label,
                score: this.state.score,
                passing: pass
            }
        ]
    }
});
```

glidej commented on src/components/GradeForm.js:
```diff
@@ -0,0 +1,82 @@
+                
+                {this.state.grades.map(data => {
+                    const passInLoop = data.score >= 12;
+                    return <Grade label={data.label} score={data.score} total={20} passing={passInLoop}/>
```
Whenever you `map` a series of components into a list like this, React needs a `key` associated with each element so it can track it. Without a key, React will be forced to re-render this List View each time `grades` changes (this list is small, so it's not a big deal, but imagine if the list was 1000s entries long!). It's a good idea to get into the habit.

A key should be a concatenated string that uniquely identifies this row _and_ it's state, so something like ``${data.label}_${data.score}_${passInLoop}`` might be a good option.

glidej commented on src/components/Inbox/Inbox.js:
```diff
@@ -12,13 +12,36 @@ class Inbox extends Component {
+  handleMarkAsRead = (index, read) => {
+    this.setState(prev => {
+      const newList = prev.messages.slice(0);
+      const newItem = Object.assign({}, prev.messages[index]);
```
The spread operator thing I mentioned elsewhere is very similar to Object.assign here. You could take advantage of assign here to merge in your new state!

```
Object.assign({}, prev.messages[index], {
    read: read
});
```

glidej commented on src/components/TimeLog/TimeForm.js:
```diff
@@ -8,19 +8,47 @@ class TimeForm extends Component {
+      end: parseTime(this.state.end)
+    }
+    this.props.onAdd(entry);
+    this.setState({
```
One thing I like to do when I have to "reset state" is create a literal that's called `INITIAL_STATE` with the properties empty/null. When I setup my `state` object, like you are on line 6, you can pass INITIAL_STATE there, and then when I need to reset state, I can pass the _same_ object here. The result is the same, but I have a central point of configuration for what "empty" looks like and that means i'm less likely to forget to update one of these as features evolve.

glidej commented on src/components/WeatherList.js:
```diff
@@ -0,0 +1,38 @@
+            <div className="WeatherList">
+                <WeatherForm onAdd={this.handleAddEntry} foo="bar"/>
+                {this.state.list.map((data, index) =>
+                    <Weather key={index} conditions={data.conditions} highF={data.highF} lowF={data.lowF} />
```
You want to base that `key` on more than just the index because React uses this key to identify if an element should be re-rendered. Because this is just the _position_ of the thing in the array, it's not the best indicator that it might need to be re-renders. More [here](https://reactjs.org/docs/lists-and-keys.html).

derekquinn commented:
@glidej Thanks for your thoughtful and helpful review here. At this point, I have addressed all of your comments except for one. I still need to play with Inbox a bit more. More React learning opportunities ahead!
